extensions/sql-database-projects/src/test/projectController.test.ts

```diff
@@ -1013,6 +1013,66 @@ describe('ProjectsController', function (): void {
 			project = await Project.openProject(project.projectFilePath);
 			should(Object.keys(project.sqlCmdVariables).length).equal(1, 'The project should only have 1 sqlcmd variable after deletion');
 		});
+
+		it('Should add sqlcmd variable', async function (): Promise<void> {
+			let project = await testUtils.createTestProject(baselines.openSdkStyleSqlProjectBaseline);
+			const sqlProjectsService = await utils.getSqlProjectsService();
+			await sqlProjectsService.openProject(project.projectFilePath);
+
+			const projController = new ProjectsController(testContext.outputChannel);
+			const projRoot = new ProjectRootTreeItem(project);
+
+			should(Object.keys(project.sqlCmdVariables).length).equal(2, 'The project should start with 2 sqlcmd variables');
+
+			const inputBoxStub = sinon.stub(vscode.window, 'showInputBox');
+			inputBoxStub.resolves('');
+			await projController.addSqlCmdVariable(createWorkspaceTreeItem(projRoot.children.find(x => x.friendlyName === constants.sqlcmdVariablesNodeName)!));
+
+			// reload project
+			project = await Project.openProject(project.projectFilePath);
+			should(Object.keys(project.sqlCmdVariables).length).equal(2, 'The project should still have 2 sqlcmd variables if no name was provided');
+
+			inputBoxStub.reset();
+			inputBoxStub.onFirstCall().resolves('newVariable');
+			inputBoxStub.onSecondCall().resolves('testValue');
+			await projController.addSqlCmdVariable(createWorkspaceTreeItem(projRoot.children.find(x => x.friendlyName === constants.sqlcmdVariablesNodeName)!));
+
+			// reload project
+			project = await Project.openProject(project.projectFilePath);
+			should(Object.keys(project.sqlCmdVariables).length).equal(3, 'The project should have 3 sqlcmd variable after adding a new one');
+		});
+
+		it('Should update sqlcmd variable', async function (): Promise<void> {
+			let project = await testUtils.createTestProject(baselines.openSdkStyleSqlProjectBaseline);
+			const sqlProjectsService = await utils.getSqlProjectsService();
+			await sqlProjectsService.openProject(project.projectFilePath);
+
+			const projController = new ProjectsController(testContext.outputChannel);
+			const projRoot = new ProjectRootTreeItem(project);
+
+			should(Object.keys(project.sqlCmdVariables).length).equal(2, 'The project should start with 2 sqlcmd variables');
+
+			const inputBoxStub = sinon.stub(vscode.window, 'showInputBox');
+			inputBoxStub.resolves('');
+			const sqlcmdVarToUpdate = projRoot.children.find(x => x.friendlyName === constants.sqlcmdVariablesNodeName)!.children[0];
+			const originalValue = project.sqlCmdVariables[sqlcmdVarToUpdate.friendlyName];
+			await projController.editSqlCmdVariable(createWorkspaceTreeItem(sqlcmdVarToUpdate));
+
+			// reload project
+			project = await Project.openProject(project.projectFilePath);
```

Comment:
Why the project reloads?  Shouldn't `project.sqlCmdVariables` be updated without needing to reload?

Reply:
The project stored in STS is updated, but the local TS representation of the project isn't updated unless the project is reloaded right now. When using ADS, the project tree gets refreshed after actions that update the project, but for tests the project needs to manually reloaded to get the updated state of the project.

Reply:
Hmm.  Perhaps any time an add/delete function in project.ts is called, it should refresh the appropriate list of items so we don't need to re-read the entire project?